### PR TITLE
Plxpr integrates well with all templates

### PR DESCRIPTION
**Pull request by dwierichs**
**To do**
- [x] Actually implement the custom binding functions
- [x] Write the custom binding tests
- [x] Write a test that asserts that all modified templates as determined automatically have their own test.

**Context:**
Introducing the plxpr capturing mechanism requires us to pay attention to operations that have special call signatures, such as templates.

**Description of the Change:**
This PR adds a test suite for templates.
For templates with special initialization signatures, a custom `_primitive_bind_call` method is added, together with dedicated tests for these custom methods.

**Benefits:**
Compatibility of Plxpr with the code stack

**Possible Drawbacks:**
More than 1 second of testing time added.
A lot of test cases
Future template additions _will_ have to write additional tests in `tests/capture/test_templates.py` in the current test suite design.

**Related GitHub Issues:**

[sc-61425]

### CODE REVIEW

File: pennylane/templates/subroutines/qdrift.py
Comment (albi3ro) [inline pennylane/templates/subroutines/qdrift.py:177]: Given `hamiltonian` should actually be in the `data`, potentially we should define `_flatten` too.

File: pennylane/templates/subroutines/hilbert_schmidt.py
Comment (albi3ro) [inline pennylane/templates/subroutines/hilbert_schmidt.py:116]:
```suggestion
    def _primitive_bind_call(cls, *params, v_function, v_wires, u_tape, id=None):
        return cls._primitive.bind(*params, v_function=v_function, v_wires=v_wires, u_tape=u_tape, id=id)
```

This might be a good use-case for writing out the call signature. Not sure how this'll work with `v_wires` being metadata here, but I don't think we need to be too concerned about the beauty of this operation.

File: pennylane/templates/subroutines/qdrift.py
Comment (albi3ro) [inline pennylane/templates/subroutines/qdrift.py:147]: Potentially could be a good separate task to make all of these "with target operations" operators inherit from `SymbolicOp`. I think that would need to be a separate task, but it would. make a lot of these registrations a lot easier, since they would just inherit the behavior from `SymbolicOp`.

File: pennylane/templates/subroutines/hilbert_schmidt.py
Comment (dwierichs) [reply pennylane/templates/subroutines/hilbert_schmidt.py:116]: Ah of course, that's so much nicer! I guess `u_tape` being metadata could also lead to problems, like `v_wires` :thinking:

File: pennylane/templates/subroutines/qdrift.py
Comment (dwierichs) [reply pennylane/templates/subroutines/qdrift.py:147]: I'll create a PR :)

**Review by albi3ro — APPROVED** [review]
Looks like there's a remaining test failure due to a regex on an error message, but happy to get this one that's fixed. Since CI will block merging anyway, I'm approving now.

File: pennylane/templates/subroutines/qdrift.py
Comment (dwierichs) [reply pennylane/templates/subroutines/qdrift.py:177]: Done.

File: pennylane/templates/subroutines/basis_rotation.py
Comment (dime10) [inline pennylane/templates/subroutines/basis_rotation.py:106]: `type` on a class returns its meta class? :)

File: pennylane/templates/subroutines/fermionic_double_excitation.py
Comment (dime10) [inline pennylane/templates/subroutines/fermionic_double_excitation.py:506]: This template has wires no?

File: pennylane/templates/subroutines/hilbert_schmidt.py
Comment (dime10) [inline pennylane/templates/subroutines/hilbert_schmidt.py:114]: Is it some specific code in the template that is not traceable with wires?

File: pennylane/templates/subroutines/qmc.py
Comment (dime10) [inline pennylane/templates/subroutines/qmc.py:346]: This op has target wires, should we not include them as the operation wires?

File: pennylane/templates/subroutines/qpe.py
Comment (dime10) [inline pennylane/templates/subroutines/qpe.py:152]: Even here I would say estimation wires could also be considered part of the target wires of the operation 🤔

**Review by dime10 — COMMENTED** [review]
Looks great! I think my main concern would be discussing why most wires are made to be constants.

File: pennylane/templates/subroutines/hilbert_schmidt.py
Comment (albi3ro) [reply pennylane/templates/subroutines/hilbert_schmidt.py:114]: Basically hilbert-schmidt is special operator that we've decided not to deal with right now.  For example, it accepts a tape. The operations in the tape might be abstarct, but the tape itself won't be.

It also accepts a function, but it doesn't really make sense to have nested jaxpr here either.

My opinion is that if we want `HilbertSchmidt` to work with catalyst and capture, we should rewrite the call signature.

File: pennylane/templates/subroutines/basis_rotation.py
Comment (albi3ro) [reply pennylane/templates/subroutines/basis_rotation.py:106]: Actually this might not be such a good idea.  `AbstractMetaABC._primitive_bind_call` just raises a `NotImplementedError`.
I think `super()._primitive_bind_call` might be what we want instead.

File: pennylane/templates/subroutines/hilbert_schmidt.py
Comment (dime10) [reply pennylane/templates/subroutines/hilbert_schmidt.py:114]: This sounds similar to the control flow operators we have, like for loop. Could the tape be turned into jaxpr and attached as metadata to the operator (same as jax and catalyst do for primitives with a "body")?

I'm happy to delay this to a future PR though.

File: pennylane/templates/subroutines/basis_rotation.py
Comment (dwierichs) [reply pennylane/templates/subroutines/basis_rotation.py:106]: Yup, it was, because there was an unrelated `xfail` taking the test out of the suite :facepalm: Good catch!

Comment (dwierichs) [reply pennylane/templates/subroutines/basis_rotation.py:106]: Turns out, it's basically not possible to support dynamic wires and matrices with this operation's signature.
Or at least I ran out of attempts and am clueless as to how adapt further.
This is related to #5521, which we decided not to fix. `BasisRotation(mat, wires=wires)` is not valid Python code.

So I fixed the issue above, switching to `super()._primitive_bind_call`, but the test is still `xfail`ing.

**Review by dwierichs — COMMENTED** [review]
@dime10 Yes you are right about the wires binding, I think.
I can try and refine this PR to respect those special settings of split wires.
Alternatively, we could go through with this PR and consider it an upgrade  to allow dynamic wires in those templates :upside_down_face:

**Comment by dime10** [comment]
> @dime10 Yes you are right about the wires binding, I think. I can try and refine this PR to respect those special settings of split wires. Alternatively, we could go through with this PR and consider it an upgrade to allow dynamic wires in those templates 🙃

That's fine by me, we can leave it up to a follow-up pr to support dynamic wire sets split across multiple arguments :)

File: pennylane/templates/subroutines/basis_rotation.py
Comment (dime10) [reply pennylane/templates/subroutines/basis_rotation.py]: Is the issue that `BasisRotation` can't be reconstructed from the jaxpr primitive because of the argument ordering? (Or is there already an issue during the construction of the primitive?)

**Review by dime10 — APPROVED** [review]

File: pennylane/templates/subroutines/basis_rotation.py
Comment (dwierichs) [reply pennylane/templates/subroutines/basis_rotation.py]: I fixed this now, by providing `BasisRotation` not only with a custom `_primitive_bind_call` but also a custom `_primitive.def_impl`, which takes the unconventional ordering of arguments into account.
This works, but I'd bet that we come back to this some later time and revert it to unify signatures.
However, as we decided to not fix #5521, it's only consistent to have the deviating signature in the capture-related code as well.

File: pennylane/templates/subroutines/hilbert_schmidt.py
Comment (dwierichs) [reply pennylane/templates/subroutines/hilbert_schmidt.py:114]: created [a story](https://app.shortcut.com/xanaduai/story/64182/plxpr-can-capture-hilbertschmidt)

**Pull request merged**
